### DISCUSSION TITLE
Transfer client configuration between environments

### DISCUSSION THREAD

**Discussion by straurob** [keycloak/keycloak#12976] [Q&A]
For securing a frontend application, I created a new Keycloak client with a custom configuration:

- mapper which includes "client roles"
- scope configuration
- client-specific roles (composite and non-composite roles)

This setup works fine in the local development setup. Now we need to transfer this configuration to the other environments like develop/preproduction/production stage.

As far as I understand, Keycloak offers the following exports:

1. Complete realm
2. Specific client

It looks as if both apporaches have some major drawbacks. Either I would need to overwrite the complete realm (which I definitely don't want to do in production) or I can import the basic client configuration which is missing all the roles.

And as soon as we, for example, add more roles later on, then we would need to re-configure all stages manually.

Is there some "good practice" how to deal with that? Does keycloak offer some kind of "sync" between stages?

**Comment by XcrigX** [comment]
@straurob  - Did you even come up with anything?   These folks wrote a Keycloak "configurer" which programmatically can create everything - custom roles, role mappers, auth flows, etc. It's pre-17.0/Quarkus but could perhaps be adapted if it doesn't work outright. 
Look at the keycloak-config sub directory:  https://github.com/Alvearie/keycloak-extensions-for-fhir

But I'm wondering if there are other ways - surely people need to reproduce complex configurations across environments..

**Reply by straurob** [reply]
Actually, I haven't worked on this topic any further and so we haven't investigated if there are better or more robust solutions.

**Reply by JeronimoMendes** [reply]
@XcrigX were you able to find a nice solution? 

I've tried to have a custom keycloak docker image built with a realm configuration, which then I can pull into different deployments, but it fails when I connect Keycloak to a database, reseting everything back to the "master" realm.

**Reply by XcrigX** [reply]
@JeronimoMendes  - Not really - I've used the referenced project above to do it. It basically uses a config file and then calls the keycloak admin cli methods to configure the realm. It works, but when things change in the admin cli between versions it takes some work to figure it out, so it's not the easiest thing to maintain. 
The alternative, AFAIK is to export a realm configuration from one server then import to another - but after you import you have to go and make manual corrections to things.

**Comment by sac-cas** [comment]
In the past I used an ansible module (https://docs.ansible.com/ansible/latest/collections/community/general/keycloak_client_module.html)  and customized it/created new ones. These modules all use the REST API of Keycloak. The more complex it is the more you have to get lots of configs from keycloak to match them with your changes.
It is the best way I know to set up a clean Keycloak with the same configuration. So you can test it locally and you can deploy it to stages. 
But you need to be careful while using the ansible states especially when using manual approvals. Sometimes you need to do 2 or more "deployments"/ansible apply when you have changed a lot and these steps must be exactly the same in any stage.
